File: app/src/main/java/com/konifar/materialcat/views/listeners/OnLoadMoreScrollListener.java
package com.konifar.materialcat.views.listeners;

import android.widget.AbsListView;

public abstract class OnLoadMoreScrollListener implements AbsListView.OnScrollListener {

    private int visibleThreshold = 5;
    private int currentPage = 0;
    private int previousTotalItemCount = 0;
    private boolean loading = true;
    private int startingPageIndex = 0;
    private AbsListView.OnScrollListener innerOnScrollListener;

    public OnLoadMoreScrollListener() {
        //
    }

    public OnLoadMoreScrollListener(AbsListView.OnScrollListener innerOnScrollListener) {
        this.innerOnScrollListener = innerOnScrollListener;
    }

    public OnLoadMoreScrollListener(int visibleThreshold) {
        this.visibleThreshold = visibleThreshold;
    }

    public OnLoadMoreScrollListener(int visibleThreshold, int startPage) {
        this.visibleThreshold = visibleThreshold;
        this.startingPageIndex = startPage;
        this.currentPage = startPage;
    }

    public int getItemCountOffset() {
        return 0;
    }

    @Override
    public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
        if (innerOnScrollListener != null) {
            innerOnScrollListener.onScroll(view, firstVisibleItem, visibleItemCount, totalItemCount);
        }

        totalItemCount -= getItemCountOffset();
        totalItemCount = Math.max(0, totalItemCount);

        if (totalItemCount < previousTotalItemCount) {
            this.currentPage = this.startingPageIndex;
            this.previousTotalItemCount = totalItemCount;
            if (totalItemCount == 0) {
                this.loading = true;
            }
        }

        if (loading && (totalItemCount > previousTotalItemCount)) {
            loading = false;
            previousTotalItemCount = totalItemCount;
            currentPage++;
        }

        if (!loading && (totalItemCount - visibleItemCount) <= (firstVisibleItem + visibleThreshold)) {
            onLoadMore(currentPage + 1, totalItemCount);
            loading = true;
        }
    }

    public abstract void onLoadMore(int page, int totalItemsCount);

    @Override
    public void onScrollStateChanged(AbsListView view, int scrollState) {
        //
    }

}

File: app/src/main/kotlin/com/konifar/materialcat/views/listeners/OnLoadMoreScrollListener.kt
package com.konifar.materialcat.views.listeners

import android.widget.AbsListView

abstract class OnLoadMoreScrollListener : AbsListView.OnScrollListener {

    private var visibleThreshold = 5
    private var currentPage = 0
    private var previousTotalItemCount = 0
    private var loading = true
    private var startingPageIndex = 0
    private var innerOnScrollListener: AbsListView.OnScrollListener? = null

    constructor() {
        //
    }

    constructor(innerOnScrollListener: AbsListView.OnScrollListener) {
        this.innerOnScrollListener = innerOnScrollListener
    }

    constructor(visibleThreshold: Int) {
        this.visibleThreshold = visibleThreshold
    }

    constructor(visibleThreshold: Int, startPage: Int) {
        this.visibleThreshold = visibleThreshold
        this.startingPageIndex = startPage
        this.currentPage = startPage
    }

    val itemCountOffset: Int
        get() = 0

    override fun onScroll(view: AbsListView, firstVisibleItem: Int, visibleItemCount: Int, totalItemCount: Int) {
        var totalItemCount = totalItemCount
        innerOnScrollListener?.onScroll(view, firstVisibleItem, visibleItemCount, totalItemCount)

        totalItemCount -= itemCountOffset
        totalItemCount = Math.max(0, totalItemCount)

        if (totalItemCount < previousTotalItemCount) {
            this.currentPage = this.startingPageIndex
            this.previousTotalItemCount = totalItemCount
            if (totalItemCount == 0) {
                this.loading = true
            }
        }

        if (loading && totalItemCount > previousTotalItemCount) {
            loading = false
            previousTotalItemCount = totalItemCount
            currentPage++
        }

        if (!loading && totalItemCount - visibleItemCount <= firstVisibleItem + visibleThreshold) {
            onLoadMore(currentPage + 1, totalItemCount)
            loading = true
        }
    }

    abstract fun onLoadMore(page: Int, totalItemsCount: Int)

    override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {
        //
    }

}
